Add support for embedding game process in the Android Editor

m4gr3d opened the pull request:
Follow up to https://github.com/godotengine/godot/pull/97257 and https://github.com/godotengine/godot/pull/99010, this PR implements game embedding for the Android editor.


https://github.com/user-attachments/assets/9a9238b9-d5e6-4bc1-9b8d-233b93c2a405


Due to differences in platform capabilities, the Android implementation uses a custom floating window, implemented using Android views, which replaces the Game workspace view seen on other platforms; i.e: when embedded, the game view on Android is always floating.
The implementation also leverages Android picture-in-picture support (which was added in https://github.com/godotengine/godot/pull/95700) to allow the game view to remain on top (when enabled) while users interact with the editor.

https://github.com/user-attachments/assets/f88479cf-6841-4363-8189-aa9ffb314c91



**Note:**
- The PR looks big, but most of the changes are in the Android editor module, so there is little risk to the Godot Android core
  - The majority of the changes in the editor module are additions of media assets to recreate the look of the game view 
- The `run/window_placement/play_window_pip_mode` editor setting has been removed, instead PiP can be explicitly triggered when running in non-embedded mode, or is implicitly used in embedded mode when `Keep on Top using PiP` is enabled.

https://github.com/user-attachments/assets/9621f8f4-61fa-43fb-b4d3-f949af891b96

### Things missing
- Ability to drag and resize the floating game view

### Limitations
- Embedded game mode is disabled for the XR Editor, support will be added in 4.5

Closes https://github.com/godotengine/godot-proposals/issues/10884

*Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/11309*

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

m4gr3d committed:
Add support for embedding game process in the Android Editor

- Implement Android editor specific `EmbeddedGodotGame` to support embedding the game window in the Android editor

syntaxerror247 commented:
I haven't reviewed the code or tested this new mode yet. However, based on the screen recording you shared, it looks like the top toolbar remains visible even when embedded mode is turned off ( while running in split-screen). I think, it would be better to disable or hide the toolbar when embedded mode is off, I takes unnecessary space.

dsnopek commented:
A couple other things I noticed in my testing:

- When the game disappears, it wasn't immediately clear to me that I had to press the "Game" tab to bring it back (since I pressed play to start the game, I never interacted with the "Game" tab). Some note about that when the game window minimizes could be helpful? Otherwise, my gut reaction was to look through the running Android apps (which it would be when not embedded) and, of course, it wasn't there.
- It's also a little weird that the "Game" tab looks selected after I press it, but whatever the previous tab that was active is what will still be there, and will still work and everything. It'd be nice if it switched back to showing 2D or 3D or whatever as selected.
- My tablet doesn't support PiP, and so the checkbox doesn't do anything. Would it be possible to not have the checkbox if PiP isn't supported? (FYI, I also tested on my phone, which does support PiP and PiP seems to be working fine.)

llama-nl commented:
The visibility toggle button doesn't seem to work. But is can be change from inspector.
https://github.com/user-attachments/assets/2266c2d0-8855-46e9-954d-b598fa35da8d

Also could you make those button little smaller. Those buttons are taking lot of space. 

And also please remove the the round corner and the bg shadow :/
![Screenshot_20250207_075912](https://github.com/user-attachments/assets/eedba1f5-e0dd-4fa7-a5ad-fbe6e296f7a8)

m4gr3d commented:
> The visibility toggle button doesn't seem to work. But is can be change from inspector.
> https://github.com/user-attachments/assets/2266c2d0-8855-46e9-954d-b598fa35da8d

@llama-nl Seems like the visibility toggle is working in the recording you took.. Are you referring to something else?

llama-nl commented:
> @llama-nl Seems like the visibility toggle is working in the recording you took.. Are you referring to something else?
The toggle button is working. But the object/mesh it still visible no matter what:


https://github.com/user-attachments/assets/d28f4f29-1840-4068-a59d-fc3a85f50b49

m4gr3d commented:
> > @llama-nl Seems like the visibility toggle is working in the recording you took.. Are you referring to something else?
> 
> @m4gr3d The toggle button is working. But the object/mesh it still visible no matter what:
> 
> 
> https://github.com/user-attachments/assets/d28f4f29-1840-4068-a59d-fc3a85f50b49
> 
> 

@llama-nl that's the expected behavior; the visibility toggle is for the selection box around the selected object, not for the object itself. 
Please refer to the feature description in https://github.com/godotengine/godot/pull/97257

syntaxerror247 commented:
I am currently testing it on my phone, and Embed mode works great! However, I’m experiencing some UI issues, which are also visible in the screen recording you shared.

https://github.com/user-attachments/assets/38bc8423-7366-4c7d-a50f-003afb9164c5

As you can see, when switching to fullscreen mode, it doesn’t fully expand and leaves a margin around the borders. Additionally, the status bar becomes visible, which I assume is unintended.

Another minor issue is that in PiP mode, a shadow appears around the borders. I’m not sure if this is intentional, but it seems out of place.

syntaxerror247 commented:
https://github.com/user-attachments/assets/08ff67b2-c677-4a00-876e-d3dff1bfede7

Running the game with Embed mode disabled tries to launch it in split screen. While this works well for tablets, it’s not optimal for smartphones and is causing issues for me. It's currently not possible for me to run the game in fullscreen.

On a regular smartphone, it should default to fullscreen mode, just like it did previously, with a button to switch to PiP mode when needed.

syntaxerror247 reviewed (changes requested):
I tested the latest changes, and they look much better now. My previous concerns about the shadow around the border in PiP mode and the margin space around the *GameWindow* in fullscreen mode is resolved now.  

However, when *embed mode* is disabled, the game still tries to open in *adjacent mode* (split-screen), which is causing the issues I mentioned earlier https://github.com/godotengine/godot/pull/102492#issuecomment-2642027354. This needs to be fixed.

syntaxerror247 commented:
I was trying to figure out why these issues occurred when running the game in `adjacent mode`. It turned out that my device's orientation was locked in portrait mode. After unlocking it, the game now switches to adjacent mode without any issues, and I can also switch to fullscreen mode using the option provided by the system. 

https://github.com/user-attachments/assets/0ae2afeb-b487-4a47-8bc7-54d3f70b9438

I would still prefer to launch it in fullscreen mode when `embed mode` is off, but I think this PR is **ready** now. We can add this option later.

m4gr3d commented on platform/android/java/editor/src/main/java/org/godotengine/editor/embed/EmbeddedGodotGame.kt:
```diff
@@ -0,0 +1,145 @@
+				if (gameMenuFragment?.isAlwaysOnTop() == true) {
+					enterPiPMode()
+				} else {
+					minimizeGameWindow()
```
A toast would help, but I'm concerned it'll get annoying fast if it shows every time the embed window is minimized.

One alternative would be to use a [snackbar](https://m2.material.io/components/snackbars) with a `disable` button, so the user can stop it from showing.. thoughts?

syntaxerror247 replied:
Yeah, `Snackbar` would work as well, but it's not part of the base Android SDK. Do you plan to add the Material Components dependency or create a custom UI?

syntaxerror247 replied:
Hey @m4gr3d, I've implemented the custom Snackbar. You can check it out here: https://github.com/syntaxerror247/godot/commit/f47d84ed1beb5a28f92bfdcd601238f7f529e8a0

syntaxerror247 replied:
Should I create a PR for it?

m4gr3d replied:
Yes!

syntaxerror247 replied:
https://github.com/godotengine/godot/pull/102590

m4gr3d commented:
@syntaxerror247 @dsnopek @llama-nl The latest version of the PR should address most of your feedback, please take another look!


> My tablet doesn't support PiP, and so the checkbox doesn't do anything. Would it be possible to not have the checkbox if PiP isn't supported? (FYI, I also tested on my phone, which does support PiP and PiP seems to be working fine.)

@dsnopek The PiP button is gated by a platform check in https://github.com/godotengine/godot/blob/ba0c4635daa2ee62968a6b869f36abdd15f30381/platform/android/java/editor/src/main/java/org/godotengine/editor/GodotGame.kt#L220 to prevent it from showing if it's not supported; it may be that your device doesn't properly implement that api. Could you grab the list of features it reports using https://android.stackexchange.com/a/82181.


> > I haven't reviewed the code or tested this new mode yet. However, based on the screen recording you shared, it looks like the top toolbar remains visible even when embedded mode is turned off ( while running in split-screen). I think, it would be better to disable or hide the toolbar when embedded mode is off, I takes unnecessary space.
> 
> I have been testing this, and I can confirm that this is indeed the case: even with embedding disabled, the toolbar still appears. I agree that it shouldn't be there in this case.

@syntaxerror247 @dsnopek Note that on desktop (tested on Windows), with embedding disabled you can still use the game menu functionality on the regular play window. I found that capability valuable on both Android and XR, but found the need to switch back & forth between windows cumbersome, hence why I added the ability for the regular play window to host the game menu.
However to address your feedback, I've added another editor setting to disable the game menu on the regular play window! Do note however that the game menu on Android also hosts the `minimize`, `enter pip` and `close` button for the regular window, so disabling it takes those away.

m4gr3d commented:
> Could you grab the list of features it reports using https://android.stackexchange.com/a/82181.

@dsnopek FYI, this is what I get on my device when I run the `adb shell pm list features|grep picture` command:
```
feature:android.software.picture_in_picture
```

syntaxerror247 commented on platform/android/java/editor/src/main/java/org/godotengine/editor/BaseGodotEditor.kt:
```diff
@@ -114,22 +121,47 @@ abstract class BaseGodotEditor : GodotActivity() {
+	private val embeddedGameStateLabel: TextView? by lazy { findViewById<TextView?>(R.id.embedded_game_state_label)?.apply {
+		setOnClickListener {
+			if (text.isNotBlank()) {
+				Toast.makeText(applicationContext, text, Toast.LENGTH_LONG).show()
```
@m4gr3d Let's remove the toast message from here. Since there's already a message displayed on screen, repeating the same information when the user clicks on it doesn't seem necessary.

syntaxerror247 replied:
What do you think about hiding the `GameWindow` entirely and only showing a toast message when the user clicks on the `Game` tab while game is not running?

llama-nl replied:
it will be better if you just put a button in game tab that run the game in embed game mode. 
![Screenshot_20250207_173022](https://github.com/user-attachments/assets/5fa0f75d-9423-4b23-9f27-e72b8f479336)

m4gr3d replied:
> What do you think about hiding the `GameWindow` entirely and only showing a toast message when the user clicks on the `Game` tab while game is not running?

@syntaxerror247 In the earlier version that would have make sense, but the newer version also provides access to the game menu bar when the game is not running which allows the user to pre-configure some settings including whether to run embedded or not, so there's value in showing it.

m4gr3d replied:
> it will be better if you just put a button in game tab that run the game in embed game mode.

@llama-nl Let me experiment with that and see what's like.

llama-nl commented:
Everything is as expected. Just I am unable to play game in `side-by-side with editor` mode with Game Embed Mode `Enabled`. Also the `enable_game_menu_bar` documentation is missing. 

And a proposal: could you make the embed window center of the screen or a way to move the window.

syntaxerror247 commented:
> Everything is as expected. Just I am unable to play game in `side-by-side with editor` mode with Game Embed Mode `Enabled`.

That's the expected behavior. You can run the game `side-by-side with editor` after disabling `Embed` Mode and keeping `enable_game_menu_bar` enabled for all games.

> And a proposal: could you make the embed window center of the screen or a way to move the window.

The ability to drag and resize the window is planned, but it requires additional work and will be added in a follow-up PR (most probably in 4.4.1).

dsnopek commented:
> @dsnopek FYI, this is what I get on my device when I run the `adb shell pm list features|grep picture` command:
> 
> ```
> feature:android.software.picture_in_picture
> ```

Ah, ok, I get that on both my phone and my tablet, even though my tablet seems unable to actually put the embedded window in PiP. So, I guess it must be some issue with my tablet.

> However to address your feedback, I've added another editor setting to disable the game menu on the regular play window! Do note however that the game menu on Android also hosts the `minimize`, `enter pip` and `close` button for the regular window, so disabling it takes those away.

Hm. This option works and does technically solve the problem, but having it as an editor setting makes it hard to get to or notice that it's there.

My main issue with the menu bar is that it can mess up the scaling of the game by taking up that extra space.

For example, here is the 2D platform game on my phone without the menu bar:

![Screenshot_20250207_091957](https://github.com/user-attachments/assets/aea6d2e6-bace-4af6-bb75-56762db6e946)

And here it is with the menu bar:

![Screenshot_20250207_091935](https://github.com/user-attachments/assets/f7c8aaa3-4eea-4193-bb3e-e028c25a13e7)

Could we maybe include this setting (and perhaps some others) in a more accessible place? Like, maybe the "Game" tab could show this setting as well as the ones about embedding and PiP, rather than just popping up the window?

Or, maybe there could be some way to hide the menu bar, sliding it away like a drawer, or something like that? Then it's not a setting but just a way of interacting with it to hide/show it.

m4gr3d commented:
> Or, maybe there could be some way to hide the menu bar, sliding it away like a drawer, or something like that? Then it's not a setting but just a way of interacting with it to hide/show it.

@dsnopek I can add a button to show / hide the menu bar. That would also allow to remove the `run/window_placement/enable_game_menu_bar` editor setting since the button would provide the behavior @syntaxerror247 and you were requesting.

syntaxerror247 commented on platform/android/java/editor/src/main/res/layout/godot_editor_layout.xml:
```diff
@@ -1,14 +1,56 @@
+
+			<TextView
+				android:id="@+id/embedded_game_state_label"
+				android:layout_width="match_parent"
```
Let's change this to `"wrap_content"`. Right now, the play button is capturing clicks across the entire horizontal width.

syntaxerror247 commented on platform/android/java/editor/src/main/java/org/godotengine/editor/BaseGodotEditor.kt:
```diff
@@ -114,22 +123,56 @@ abstract class BaseGodotEditor : GodotActivity() {
+
+		private const val GAME_WORKSPACE = "Game"
+
+		internal const val SNACKBAR_SHOW_DURATION_MS = 3000L
```
Let's also increase it to 5 seconds. 3 seconds works, but it feels a bit too fast to me.

syntaxerror247 replied:
Nevermind, I think 3 is fine.

m4gr3d commented:
@dsnopek @llama-nl @syntaxerror247 All of your feedback have been addressed! Can you give the latest version a try?

> When GameWindow is visible and you close the app to background and then open from recents screen it would only open Embedded Game activity, Editor would be hidden.

@syntaxerror247 That issue should also be resolved in the latest version!

syntaxerror247 commented:
Everything works great! 🎉

I've actually been testing and reviewing every change you pushed to make sure nothing slipped through, especially as we approach the end of beta.

syntaxerror247 commented on platform/android/java/editor/src/main/java/org/godotengine/editor/BaseGodotGame.kt:
```diff
@@ -0,0 +1,104 @@
+	protected open fun getProjectPermissionsToEnable(): MutableList<String> {
+		val permissionsToEnable = mutableListOf<String>()
+
+		// Check for RECORD_AUDIO permission
```
```suggestion
		// Check for RECORD_AUDIO permission.
```

syntaxerror247 commented on platform/android/java/editor/src/main/java/org/godotengine/editor/embed/EmbeddedGodotGame.kt:
```diff
@@ -0,0 +1,147 @@
+import org.godotengine.godot.utils.GameMenuUtils
+
+/**
+ * Host the Godot game from the editor when the embedded mode is enabled
```
```suggestion
 * Host the Godot game from the editor when the embedded mode is enabled.
```

syntaxerror247 commented on platform/android/java/editor/src/main/java/org/godotengine/editor/embed/EmbeddedGodotGame.kt:
```diff
@@ -0,0 +1,147 @@
+
+			MotionEvent.ACTION_MOVE -> {
+//				val layoutParams = window.attributes
+				// TODO: Add logic to move the embedded window
```
```suggestion
				// TODO: Add logic to move the embedded window.
```

syntaxerror247 commented on platform/android/java/editor/src/main/java/org/godotengine/editor/embed/EmbeddedGodotGame.kt:
```diff
@@ -0,0 +1,147 @@
+
+	override fun onPictureInPictureModeChanged(isInPictureInPictureMode: Boolean) {
+		super.onPictureInPictureModeChanged(isInPictureInPictureMode)
+		// Maximize the dimensions when entering PiP so the window fills the full PiP bounds
```
```suggestion
		// Maximize the dimensions when entering PiP so the window fills the full PiP bounds.
```

syntaxerror247 commented on platform/android/java/editor/src/main/java/org/godotengine/editor/embed/GameMenuFragment.kt:
```diff
@@ -0,0 +1,480 @@
+	}
+
+	/**
+	 * Used to be notified of events fired when interacting with the game menu
```
```suggestion
	 * Used to be notified of events fired when interacting with the game menu.
```

syntaxerror247 commented on platform/android/java/editor/src/main/java/org/godotengine/editor/embed/GameMenuFragment.kt:
```diff
@@ -0,0 +1,480 @@
+	interface GameMenuListener {
+
+		/**
+		 * Kotlin representation of the RuntimeNodeSelect::SelectMode enum in 'scene/debugger/scene_debugger.h'
```
```suggestion
		 * Kotlin representation of the RuntimeNodeSelect::SelectMode enum in 'scene/debugger/scene_debugger.h'.
```

syntaxerror247 commented on platform/android/java/editor/src/main/java/org/godotengine/editor/embed/GameMenuFragment.kt:
```diff
@@ -0,0 +1,480 @@
+		}
+
+		/**
+		 * Kotlin representation of the RuntimeNodeSelect::NodeType enum in 'scene/debugger/scene_debugger.h'
```
```suggestion
		 * Kotlin representation of the RuntimeNodeSelect::NodeType enum in 'scene/debugger/scene_debugger.h'.
```

syntaxerror247 commented on platform/android/java/editor/src/main/java/org/godotengine/editor/embed/GameMenuFragment.kt:
```diff
@@ -0,0 +1,480 @@
+		}
+
+		/**
+		 * Kotlin representation of the EditorDebuggerNode::CameraOverride in 'editor/debugger/editor_debugger_node.h'
```
```suggestion
		 * Kotlin representation of the EditorDebuggerNode::CameraOverride in 'editor/debugger/editor_debugger_node.h'.
```

syntaxerror247 commented on platform/android/java/lib/src/org/godotengine/godot/utils/GameMenuUtils.kt:
```diff
@@ -0,0 +1,117 @@
+	private val TAG = GameMenuUtils::class.java.simpleName
+
+	/**
+	 * Enum representing the "run/window_placement/game_embed_mode" editor settings
```
```suggestion
	 * Enum representing the "run/window_placement/game_embed_mode" editor settings.
```

syntaxerror247 reviewed (commented):
I spotted a bunch of missing `.`  I don't mind, but @AThousandShips wouldn’t like it 😀

llama-nl commented:
Wow, now everything is perfect. Really grate work

dsnopek commented:
> I can add a button to show / hide the menu bar.

I just tested this, and it works great! Could we have it even when in embedded mode? Because then I really don't think I'd have a reason to _not_ use embedded mode, especially if the menu bar started out hidden. Like 99% of the time, I don't think I'd want to use the buttons on the menu bar and just want to test the game, so it makes more sense to me to have it hidden and then show it, rather than the other way around. I know it's shown by default on other platforms, but on mobile screen space is more precious.

Anyway, that's not worth getting hung up on now. :-) I haven't reviewed the code very deeply, but overall this is working great in my testing!

Repiteo commented:
Thanks!
